docs/adr/0016-use-graphql-and-apollo-for-settings-redesign.md

```diff
@@ -0,0 +1,136 @@
+
+"Apollo" in this document refers to the "Apollo client" and "Apollo server" pieces of the Apollo platform¹, which can be described as a unified data layer that enables applications to interact with data from data stores and APIs. In other words, it allows us to write and handle GraphQL on the client and server. Apollo also gives us many tools out of the box like caching.
+
+This ADR serves to lay out pros and cons of using GraphQL and Apollo in the Settings Redesign project as an alternative to hitting our conventional REST endpoints.
+
+¹Apollo also offers Apollo Graph Manger and Apollo Federation which are paid services, [read more from their docs](https://www.apollographql.com/docs/intro/platform/). We do not need to use these to use GQL with Apollo server or Apollo client.
+
+## Decision Drivers
+
+- Performance
+- Easy API setup and React integration
+- Clarity around "what's happening?" and data received from the server
+- Development speed, both around initial setup and as new features roll out
+
+## Considered Options
+
+- Option A - Send standard CRUD requests to our REST endpoints
+- Option B - Layer GraphQL and Apollo on top of our REST architecture, making direct database calls for basic operations
+- Option C - Full GraphQL and Apollo integration with direct database calls
+
+## Decision Outcome
+
+Chosen option: "B - Layer GraphQL on top of our REST architecture", because:
+
+- GQL seems to be the ideal solution to the heavy number of network requests on the main Settings page where around 17 different requests are sent to our REST API, and a lot of data received is superfluous. GQL moves us towards a performant solution that consolidates these initial requests and reduces overfetching in other requests.
+- Allowing developers to query for and expect only exactly what is needed with end-to-end typing and a more "declarative" way of thinking towards data requirements makes it painfully clear what's being sent and received between client and server.
+- The React Settings application will only have to manage interacting with a single endpoint and Apollo Client makes managing the data in our React app straightforward.
+- Does not preclude option C, as we can replace or supplement direct calls to the FxA auth-server down the line. This is faster for initial development, and this option over option C will also help mitigate risk around using a relatively novel piece of technology for FxA.
+
+## Pros and Cons of the Options
+
+### Option A - Send standard CRUD requests to our REST endpoints
+
+#### Description
+
+- Don't use GraphQL. Send all requests to their corresponding endpoints.
+
+##### Pros:
+
+- No novel tech.
+- No additional setup time.
+- No tech debt.
+- Our REST API setup has been vetted by years of use in production.
+
+##### Cons:
+
+- A lack of needed performance gains -- we won't solve our heavy initial network request problem and we'll continue to overfetch data as REST does.
```

Comment:
> A lack of needed performance gains

I am curious how much of a performance gain we get from reducing the network requests. The request are pretty light and return quickly.

Reply:
Yeah, and I'm also curious how much gain we could get from HTTP/2.

Reply:
Our requests in many cases are already using HTTP/2, as the load-balancers now support that we use. HTTP/2 only helps with the request setup, the problem here is that the requests are done sequentially. The requests could definitely be optimized manually in the current setup, perhaps even transparently to the app (by using a network intercept hook to cache the call if it's done more than once per 2-seconds for example). We could also fire all the requests at once, rather than sequentially, but none of this is efficient or great for our server load.

I believe graphql offers a winning proposition here, because it allows components in the page to indicate the data they need, which can be automatically collapsed by the apollo tooling and expressed as a single graphql request. Doing a similar thing for RESTful calls requires manual optimization every time the settings page changes, to ensure we're doing as little calls as needed for the data fetched. The server then benefits as well since it can do the minimum work required to get the data needed.

Reply:
Oops, to be more specific to this point, perhaps a more accurate phrasing would be that:
- REST results in us either overfetching data, doing more queries than needed, or spending additional time manually optimizing calls as we update the settings page.

Reply:
Heh, I wasn't suggesting HTTP/2 would _fix_ how the requests are done, but curious how much of a gain it would be.

> Our requests in many cases are already using HTTP/2

This is not true for content server, auth, or profile.  Payments is using HTTP/2 though.

Reply:
> the problem here is that the requests are done sequentially

This is a good point, and perhaps should be made explicit in the ADR.